File: solutions/ocaml/bob/2/bob.ml
open Base

let has_letters s = String.exists ~f:Char.is_alpha s
let is_yelling s = String.(uppercase s = s) && has_letters s
let is_question s = String.is_suffix ~suffix:"?" s
let is_silence s = String.is_empty s

let response_for input =
  let s = input |> String.strip in
  if is_silence s then "Fine. Be that way!"
  else if is_question s && is_yelling s then "Calm down, I know what I'm doing!"
  else if is_question s then "Sure."
  else if is_yelling s then "Whoa, chill out!"
  else "Whatever."

File: solutions/clojure/bob/2/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(def silence? str/blank?)
(def question? #(str/ends-with? % "?"))
(def contains-letters? #(re-find #"[A-Za-z]" %))
(def all-uppercase? #(= % (str/upper-case %)))
(def yelling? (every-pred contains-letters? all-uppercase?))
(def yelling-question? (every-pred yelling? question?))

(defn response-for [s]
  (condp apply [(str/trim s)]
    silence? "Fine. Be that way!"
    yelling-question? "Calm down, I know what I'm doing!"
    yelling? "Whoa, chill out!"
    question?  "Sure."
    "Whatever."))